google-cloud-debugger/lib/google/cloud/debugger/transmitter.rb

```diff
@@ -13,127 +13,184 @@
+        #   @return [Array<Google::Cloud::Debugger::Breakpoint>] The
+        #   individual error event that was not submitted to Stackdriver
+        #   Debugger service.
+        attr_reader :breakpoint
+
+        def initialize message, breakpoint = nil
+          super(message)
+          @breakpoint = breakpoint
+        end
+      end
+
       ##
       # # Transmitter
       #
       # Responsible for submit evaluated breakpoints back to Stackdriver
-      # Debugger service asynchronously. It has it's own dedicated child thread,
-      # so it doesn't interfere with main Ruby application threads or the
-      # debugger agent's thread.
+      # Debugger service asynchronously. It maintains a thread pool.
       #
       # The transmitter is controlled by the debugger agent it belongs to.
-      # Debugger agent submits evaluated breakpoint into a thread safe queue,
-      # and the transmitter operates directly on the queue along with minimal
-      # interaction with the rest debugger agent components
+      # Debugger agent submits evaluated breakpoint asynchronously, and the
+      # transmitter submits the breakpoints to Stackdriver Debugger service.
       #
       class Transmitter
-        include Stackdriver::Core::AsyncActor
-
-        ##
-        # @private Default maximum backlog size for the job queue
-        DEFAULT_MAX_QUEUE_SIZE = 1000
-
         ##
-        # @private The gRPC Service object.
-        attr_accessor :service
+        # @private The gRPC Service object and thread pool.
+        attr_reader :service, :thread_pool
 
         ##
         # The debugger agent this transmiter belongs to
         # @return [Google::Cloud::Debugger::Agent]
         attr_accessor :agent
 
         ##
-        # Maxium backlog size for this transmitter's queue
-        attr_accessor :max_queue_size
+        # Maximum backlog size for this transmitter's queue
+        attr_accessor :max_queue
+        alias max_queue_size  max_queue
+        alias max_queue_size= max_queue=
+
+        ##
+        # Maximum threads used in the thread pool
+        attr_accessor :threads
 
         ##
-        # @private Construct a new instance of Tramsnitter
-        def initialize agent, service, max_queue_size = DEFAULT_MAX_QUEUE_SIZE
-          super()
+        # @private Creates a new Transmitter instance.
+        def initialize agent, service, max_queue: 1000, threads: 10
+          @agent   = agent
           @service = service
-          @agent = agent
-          @max_queue_size = max_queue_size
-          @queue = []
-          @queue_resource = new_cond
+
+          @max_queue = max_queue
+          @threads   = threads
+
+          @thread_pool = Concurrent::CachedThreadPool.new max_threads: @threads,
+                                                          max_queue: @max_queue
+
+          @error_callbacks = []
+
+          # Make sure all queued calls are completed when process exits.
+          # at_exit { stop }
```

Comment:
Why is this line commented out?

Reply:
Because of the brain damage from banging my head against this change over and over and over and over and over and over. Fixed.